hw/ip/otbn/dv/otbnsim/sim/stats.py

```diff
@@ -0,0 +1,302 @@
+    FREQ_MHZ = 100
+
+    def __init__(self, stats: ExecutionStats, elf_file_path: str):
+        self._elf_file = ELFFile(open(elf_file_path, 'rb'))
+        self._stats = stats
+        self._addr_symbol_map = _get_addr_symbol_map(self._elf_file)
+
+    def _describe_imem_addr(self, address: int) -> str:
+        symbol_name = None
+        if address in self._addr_symbol_map:
+            symbol_name = self._addr_symbol_map[address]
+        else:
+            # |func_addr| is the largest possible |sym_addr| which is at most
+            # |address|.
+            func_addr = 0
+            for sym_addr in self._addr_symbol_map.keys():
+                if sym_addr <= address and sym_addr > func_addr:
+                    func_addr = sym_addr
+            symbol_name = self._addr_symbol_map[func_addr] + f"+{address - func_addr:#x}"
+
+        file_line = None
+        if self._elf_file.has_dwarf_info():
+            dwarf_info = self._elf_file.get_dwarf_info()
+            file_line = _dwarf_decode_file_line(dwarf_info, address)
+
+        add_info = []
+        if symbol_name:
+            add_info.append(f"{symbol_name}")
+        if file_line:
+            add_info.append(f"at {file_line[0]}:{file_line[1]}")
+
+        str = f"{address:#x}"
+        if add_info:
+            str += ' (' + ' '.join(add_info) + ')'
+        return str
+
+    def dump(self) -> str:
+        out = ""
+        out += "\n"
+        out += "Execution time\n"
+        out += "--------------\n"
+        out += self._dump_execution_time()
+        out += "\n\n"
+        out += "Instruction frequencies\n"
+        out += "-----------------------\n"
+        out += self._dump_instruction_histo()
+        out += "\n\n"
+        out += "Function call statistics\n"
+        out += "------------------------\n"
+        out += self._dump_function_call_stats()
+        out += "\n\n"
+        out += "Loop statistics\n"
+        out += "---------------\n"
+        out += self._dump_loop_stats()
+        out += "\n"
+
+        return out
+
+    def _dump_execution_time(self) -> str:
+        insn_count = self._stats.insn_count
+        stall_count = self._stats.stall_count
+        stall_percent = stall_count / insn_count * 100
+        cycles = insn_count + stall_count
+        time_ms = cycles / (self.FREQ_MHZ * 1e6) * 1e3
+
+        out = f"OTBN executed {insn_count} instructions in {cycles} cycles.\n"
+
+        out += f"The execution stalled for {stall_count} cycles "
+        out += f"({stall_percent:.01f} percent).\n"
+
+        out += f"The execution would take {time_ms:.02f} ms "
+        out += f"@ {self.FREQ_MHZ} MHz.\n"
+        return out
+
+    def _dump_instruction_histo(self) -> str:
+        return tabulate(self._stats.instruction_histo.most_common(),
+                        headers=['instruction', 'count']) + "\n"
+
+    def _dump_function_call_stats(self) -> str:
+        """ Dump function call statistics """
+        out = ""
```

Comment:
Not that it really matters, but this might be a bit more efficient if we collected up a list of strings and then collapsed them at the end with `''.join(out)`. That way, we don't have to keep appending to (immutable) string objects. It also makes the code nicer, I think:
```python3
out += [f"Function {func}\n",
        "  is called from the following functions\n"]
```
etc.

Actually, maybe the accumulator should be called `lines` and joined together with `'\n'.join(lines)` at the end.

Reply:
I played around with a number of options, and I like none of them. I therefore prefer to leave the code as-is.

And I always hated the "let's assemble lists and then use join" approach, it feels like something to please the language at the cost of readability where it could just optimize for that case. (Yes, the difference is measurable in microbenchmarks, but doesn't matter here.)

Reply:
Hmm, maybe. I think it's not too bad when we're assembling a list of lines: at least then it feels more like we can describe what we're collecting. But no strong feelings here.